common/src/main/java/com/wynntils/core/net/Download.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.net;
+
+import com.wynntils.core.WynntilsMod;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.http.HttpHeaders;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
+import org.apache.commons.io.FileUtils;
+
+public class Download {
+    static Map<HttpRequest, CompletableFuture<HttpResponse<Path>>> downloadFutures = new HashMap<>();
+    static Map<HttpRequest, CompletableFuture<Void>> processFutures = new HashMap<>();
+
+    HttpRequest request;
+    private final File localFile;
+
+    // Saved since we might need to get timestamps from the HttpResponse
+    CompletableFuture<HttpResponse<Path>> future;
+
+    public Download(File localFile) {
+        this.localFile = localFile;
+    }
+
+    public Download(File localFile, HttpRequest request) {
+        this.localFile = localFile;
+        this.request = request;
+    }
+
+    protected CompletableFuture<HttpResponse<Path>> cacheHttpResponseAsync() {
+        FileUtils.deleteQuietly(localFile);
+        localFile.getParentFile().mkdirs();
+        future = NetManager.HTTP_CLIENT
+                .sendAsync(request, HttpResponse.BodyHandlers.ofFile(localFile.toPath()))
+                .whenComplete((ignored, exc) -> {
+                    downloadFutures.remove(request);
+                });
+        // in case of failure:
+        //        FileUtils.deleteQuietly(cacheFile);
+        downloadFutures.put(request, future);
+        return future;
+    }
+
+    protected CompletableFuture<InputStream> cacheInputStreamAsync() {
+        if (request == null) {
+            try {
+                InputStream inputStream = new FileInputStream(localFile);
```

Comment:
Qodana might be right, use try with resource?

Reply:
How does that mix with future lambdas? 🤔